### ISSUE TITLE
Spurious nans found when using jax_debug_nans.

### ISSUE THREAD

**Issue by MaxiLechner** [google/jax#5698]
I am currently trying to figure out why the gradient of an energy minimization routine in jax-md returns nans by using jax_debug_nans but I am getting some funny errors when doing that.

This scan includes evaluating a force which is the gradient of an energy function.
```python
force_fn = grad(energy_fn)
```
Now ```force_fn(x)``` produces nans if and only if jax_debug_nans is set to True.  ```jit(force_fn)(x)``` on the other hand always works out fine. 

I have tried to extract the relevant parts from jax-md into a colab notebook.
https://colab.research.google.com/drive/1CYtaYI7N0vOcLN_O1RrnrspOKvlKK9Z2?usp=sharing

I believe the nans I am getting in my scan are unrelated to this issue here.

**Comment by skye** [comment]
It appears that `force_fn` is producing NaNs internally, but not in a way that affects the final output. I didn't look at your computation in detail to figure out how exactly this is the case, so not sure if this is expected or not, or if the internal NaN is affecting your scan.

Maybe it'd be useful in this case to log all NaNs without halting execution? You'll probably get a lot of logging, but if you start at the top maybe you can figure out what's going on. Probably the quickest way to do this is to install jax from github and make the modification yourself.

To install from github:
```
git clone https://github.com/google/jax
cd jax
pip install --upgrade -e .
```
Then modify [xla.py:385](https://github.com/google/jax/blob/master/jax/interpreters/xla.py#L385) to print instead of raising an exception, something like:
```python
    if FLAGS.jax_debug_nans and np.any(np.isnan(buf.to_py())):
      print(f"invalid value (nan) encountered in {name}")
      import traceback; traceback.print_stack()
      return
```

Hope this helps!

**Comment by MaxiLechner** [comment]
If I do that I get the following traceback, which doesn't seem to tell me where the issue lies.

```python
invalid value (nan) encountered in mul
  File "/Users/Maxi/IST/maxi/nan.py", line 69, in <module>
    print("force: ", force_fn(R))
  File "/Users/Maxi/IST/maxi/jax/jax/_src/traceback_util.py", line 139, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/Maxi/IST/maxi/jax/jax/api.py", line 742, in grad_f
    _, g = value_and_grad_f(*args, **kwargs)
  File "/Users/Maxi/IST/maxi/jax/jax/_src/traceback_util.py", line 139, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/Maxi/IST/maxi/jax/jax/api.py", line 811, in value_and_grad_f
    g = vjp_py(np.ones((), dtype=dtype))
  File "/Users/Maxi/IST/maxi/jax/jax/api.py", line 1798, in _vjp_pullback_wrapper
    ans = fun(*args)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/ad.py", line 121, in unbound_vjp
    arg_cts = backward_pass(jaxpr, consts, dummy_args, cts)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/ad.py", line 228, in backward_pass
    **eqn.params)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/ad.py", line 497, in bilinear_transpose
    out = lhs_rule(cotangent, y, **kwargs)
  File "/Users/Maxi/IST/maxi/jax/jax/_src/lax/lax.py", line 349, in mul
    return mul_p.bind(x, y)
  File "/Users/Maxi/IST/maxi/jax/jax/core.py", line 282, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "/Users/Maxi/IST/maxi/jax/jax/core.py", line 628, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/xla.py", line 319, in apply_primitive
    return compiled_fun(*args)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/xla.py", line 462, in _execute_compiled_primitive
    check_special(prim, out_bufs)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/xla.py", line 485, in check_special
    _check_special(prim.name, buf.xla_shape(), buf)
  File "/Users/Maxi/IST/maxi/jax/jax/interpreters/xla.py", line 495, in _check_special
    traceback.print_stack()
```

I have taken another look at it and I am pretty sure that the issue is caused by either the distance function, or more likely the safe_mask function because if I jit either of them no nans are found. There is a long thread here #1052 about the perils of differentiating jnp.where but my safe_mask function should prevent exactly that.

Having said that I am always going to jit my energy and my force function so this should not be an issue in practice. I am simply wondering what exactly is going wrong.

**Comment by skye** [comment]
@mattjj do you have any ideas?

**Comment by hawkinsp** [comment]
I'd guess this is working as intended. `jax_debug_nans` reports if *any* NaNs are produced as outputs from a computation. It doesn't have a way to distinguish between benign NaNs that you later mask out and NaNs that are dangerous. It's a debugging tool, and has its limitations!

Since the nan-debugging only ever looks at the outputs of a computation, if you use a `jit` to "enclose" the NaNs and their masking, then the NaN debugging tool won't detect them.

**Comment by MaxiLechner** [comment]
Thanks for the quick answer.
